commands/login.go

```diff
@@ -413,3 +399,149 @@ func IdentityString(pkt pktoken.PKToken) (string, error) {
+		providerConfig.Scopes = []string{"openid", "profile", "email"}
+	}
+
+	if strings.HasPrefix(providerConfig.Issuer, "https://accounts.google.com") {
+		// The Google OP is strange in that it requires a client secret even if this is a public OIDC App.
+		// Despite its name the Google OP client secret is a public value.
+		if providerConfig.ClientSecret == "" {
+			if hasAlias {
+				return ProviderConfig{}, fmt.Errorf("Error: Invalid provider argument format. Expected format for google: <alias>,<issuer>,<client_id>,<client_secret> \n")
+			} else {
+				return ProviderConfig{}, fmt.Errorf("Error: Invalid provider argument format. Expected format for google: <issuer>,<client_id>,<client_secret> \n")
+			}
+
+		}
+	}
+	return providerConfig, nil
+}
+
+// Function to create the provider from the config
+func NewProviderFromConfig(config ProviderConfig) (client.OpenIdProvider, error) {
+
+	fmt.Printf("%+v\n", config)
+	if config.Issuer == "" {
+		return nil, fmt.Errorf("Error: Invalid provider issuer value got (%s) \n", config.Issuer)
+	}
+
+	if !strings.HasPrefix(config.Issuer, "https://") {
+		return nil, fmt.Errorf("invalid provider issuer value. Expected issuer to start with 'https://' got (%s) \n", config.Issuer)
+	}
+
+	if config.ClientID == "" {
+		return nil, fmt.Errorf("invalid provider client-ID value got (%s) \n", config.ClientID)
+	}
+	var provider client.OpenIdProvider
+
+	if strings.HasPrefix(config.Issuer, "https://accounts.google.com") {
+		opts := providers.GetDefaultGoogleOpOptions()
+		opts.Issuer = config.Issuer
+		opts.ClientID = config.ClientID
+		opts.ClientSecret = config.ClientSecret
+		opts.GQSign = false
+		provider = providers.NewGoogleOpWithOptions(opts)
+	} else if strings.HasPrefix(config.Issuer, "https://login.microsoftonline.com") {
+		opts := providers.GetDefaultAzureOpOptions()
+		opts.Issuer = config.Issuer
+		opts.ClientID = config.ClientSecret
+		opts.GQSign = false
+		provider = providers.NewAzureOpWithOptions(opts)
+	} else if strings.HasPrefix(config.Issuer, "https://gitlab.com") {
+		opts := providers.GetDefaultGitlabOpOptions()
+		opts.Issuer = config.Issuer
+		opts.ClientID = config.ClientSecret
+		opts.GQSign = false
+		provider = providers.NewGitlabOpWithOptions(opts)
+	} else {
+		// Generic provider - Need signing, no encryption
+		opts := providers.GetDefaultGoogleOpOptions()
+		opts.Issuer = config.Issuer
+		opts.ClientID = config.ClientID
+		opts.GQSign = false
+		opts.ClientSecret = config.ClientSecret
+		opts.Scopes = config.Scopes
+
+		provider = providers.NewGoogleOpWithOptions(opts)
+	}
+
+	fmt.Printf("%+v\n", provider)
+
+	return provider, nil
+}
+
+// Function to retrieve the config from the env variables
+// OPKSSH_DEFAULT can be set to an alias
+// OPKSSH_PROVIDERS is a ; separated list of providers of the format <alias>,<issuer>,<client_id>,<client_secret>,<scopes>;<alias>,<issuer>,<client_id>,<client_secret>,<scopes>
+func GetProvidersConfigFromEnv() (map[string]ProviderConfig, error) {
```

Comment:
This function would be more flexible if it returns `(map[string]ProviderConfig, error)`

- If no providers are found in env vars: value=empty but not nil, err = nil
- If providers are found in env vars but you hit a validation error: value=nil, err = error
- If providers are found in env vars and they validate: value=providers found, err = nil

Then do the failback to the default providers outside of this function. That way we can tell if the providers actually came from the env vars or the default providers